File: web/react-productos/src/services/productoService.ts
import { api } from "./api";
import type { Producto, CrearProductoDTO } from "../types/producto";

export const productoService = {
  async obtenerTodos(): Promise<Producto[]> {
    const { data } = await api.get<Producto[]>("/productos");
    return data;
  },

  async obtenerPorId(id: string): Promise<Producto> {
    const { data } = await api.get<Producto>(`/productos/${id}`);
    return data;
  },

  async crear(dto: CrearProductoDTO): Promise<Producto> {
    const { data } = await api.post<Producto>("/productos", dto);
    return data;
  },

  async actualizar(id: string, dto: CrearProductoDTO): Promise<void> {
    await api.put(`/productos/${id}`, dto);
  },

  async eliminar(id: string): Promise<void> {
    await api.delete(`/productos/${id}`);
  },
};

File: api/API.Productos/API.Productos/Services/ProductoService.cs
using API.Productos.Entities;
using API.Productos.Repositories;
using System.Xml.Linq;

namespace API.Productos.Services
{
    public class ProductoService: IProductoService
    {
        private readonly IProductosRepository _repository;
        public ProductoService(IProductosRepository repository)
        {
            _repository = repository;
        }
        public IEnumerable<Producto> ObtenerTodos()
        {
            return _repository.ObtenerTodos();
        }
        public Producto? ObtenerPorId(Guid id)
        {
            return _repository.ObtenerPorId(id);
        }
        public Producto Crear(string nombre, decimal precio)
        {
            if (string.IsNullOrWhiteSpace(nombre))
                throw new ArgumentException("El nombre es requerido");

            if (precio <= 0)
                throw new ArgumentException("El precio no puede ser menor a cero");

            var producto = new Producto { Nombre = nombre, Precio = precio };
            return _repository.Agregar(producto);
        }
        public Producto? Actualizar(Guid id, string nombre, decimal precio)
        {
            if (string.IsNullOrWhiteSpace(nombre))
                throw new ArgumentException("El nombre es requerido");
            if (precio <= 0)
                throw new ArgumentException("El precio no puede ser menor a cero");

            var producto = new Producto { Id = id, Nombre = nombre, Precio = precio };
            return _repository.Actualizar(producto);
        }

        public bool Eliminar(Guid id)
        {
            return _repository.Eliminar(id);
        }
    }
}
